json.go

```diff
@@ -27,14 +30,67 @@ import (
 // CanonicalJSON re-encodes the JSON in a canonical encoding. The encoding is
 // the shortest possible encoding using integer values with sorted object keys.
 // https://matrix.org/docs/spec/server_server/unstable.html#canonical-json
+// Returns a gomatrixserverlib.BadJSONError if JSON validation fails.
 func CanonicalJSON(input []byte) ([]byte, error) {
 	if !gjson.Valid(string(input)) {
-		return nil, fmt.Errorf("invalid json")
+		return nil, BadJSONError{errors.New("gjson validation failed")}
 	}
 
 	return CanonicalJSONAssumeValid(input), nil
 }
 
+// Returns a gomatrixserverlib.BadJSONError if the canonical JSON fails enforced
+// checks or if JSON validation fails.
+func EnforcedCanonicalJSON(input []byte, roomVersion RoomVersion) ([]byte, error) {
+	if enforce, err := roomVersion.EnforceCanonicalJSON(); err == nil && enforce {
+		if err = verifyEnforcedCanonicalJSON(input); err != nil {
+			return nil, BadJSONError{err}
+		}
+	}
+
+	return CanonicalJSON(input)
+}
+
+func verifyEnforcedCanonicalJSON(input []byte) error {
+	var j interface{}
+	if err := json.Unmarshal(input, &j); err != nil {
+		return err
+	}
+	var walk func(v reflect.Value) error
+	walk = func(v reflect.Value) error {
+		for v.Kind() == reflect.Ptr || v.Kind() == reflect.Interface {
+			v = v.Elem()
+		}
+		switch v.Kind() {
+		case reflect.Array, reflect.Slice:
+			for i := 0; i < v.Len(); i++ {
+				if err := walk(v.Index(i)); err != nil {
+					return err
+				}
+			}
+		case reflect.Map:
+			for _, k := range v.MapKeys() {
+				if err := walk(v.MapIndex(k)); err != nil {
+					return err
+				}
+			}
+		case reflect.Int, reflect.Int64:
+			if v.Int() < -9007199254740991 || v.Int() > 9007199254740991 {
+				return fmt.Errorf("%d is outside of safe range", v.Int())
+			}
+		case reflect.Float64:
+			if v.Float() < -9007199254740991 || v.Float() > 9007199254740991 {
```

Comment:
https://matrix.org/docs/spec/appendices#canonical-json indicates that floats are illegal.